USB Redirection with usbredir

victortoso opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
This PR allows an user to redirect USB devices from the client machine to the VMI using [usbredir](https://gitlab.freedesktop.org/spice/usbredir/) (protocol, library and tools). The library is used in both QEMU and client side with usbredirect binary, introduced in [0.9.0 release](https://gitlab.freedesktop.org/spice/usbredir/-/releases/usbredir-0.9.0).

usbredir main dependency is libusb which works well in Linux, Windows and macos. The usbredirect binary has an addition dependency with GLib and it was tested in Windows as well as in Linux.

This PR implements a usbredir command under virtctl to call usbredirect binary to redirect a local USB devices to the remote VMI.

The current API is limited only to enable USB devices in the VMI. The user needs to add an empty structure to clientPassthrough under devices, like:
```yaml
 spec:
   domain:
     devices:
      clientPassthrough: {}
```

```bash
sudo ./cluster-up/kubectl.sh usbredir 145f:019f vmi-alpine-efi
```

You should see in the VM's dmesg:
```
[ 1169.202733] usb 1-1: USB disconnect, device number 2
[ 1180.630968] usb 1-1: new high-speed USB device number 4 using xhci_hcd
[ 1180.880803] usb 1-1: New USB device found, idVendor=145f, idProduct=019f
[ 1180.880811] usb 1-1: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[ 1180.880816] usb 1-1: Product: Trust Webcam
[ 1180.880819] usb 1-1: Manufacturer: Novatek
```

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add support to USB Redirection with usbredir
```

danielBelenky commented on pkg/virt-launcher/virtwrap/api/converter_test.go:
```diff
@@ -1333,6 +1333,15 @@ var _ = Describe("Converter", func() {
 			Expect(domain.Spec.Devices.Inputs[0].Bus).To(Equal("usb"), "Expect usb bus")
 		})
 
+		It("should enable usbredir when numberOfDevices > 0", func() {
```
Could you please also add tests for when numberOfDevices is negative and zero?

victortoso replied:
I've added those, thanks!

danielBelenky commented on cluster-up/kubectl.sh:
```diff
@@ -30,7 +30,7 @@ source ${KUBEVIRTCI_PATH}/hack/common.sh
 source ${KUBEVIRTCI_PATH}/hack/config.sh
 
-if [ "$1" == "console" ] || [ "$1" == "vnc" ] || [ "$1" == "start" ] || [ "$1" == "stop" ] || [ "$1" == "migrate" ]; then
+if [ "$1" == "console" ] || [ "$1" == "vnc" ] || [ "$1" == "usbredir" ] || [ "$1" == "start" ] || [ "$1" == "stop" ] || [ "$1" == "migrate" ]; then
```
This should be done in `kubevirt/kubevirtci`. The entire `cluster-up/` directory is imported from there.
https://github.com/kubevirt/kubevirtci/

victortoso replied:
Ah, right. I've been using this locally for testing, didn't know it cames from kubevirtci, thanks!

danielBelenky commented on pkg/virt-handler/rest/console.go:
```diff
@@ -45,17 +46,75 @@ type ConsoleHandler struct {
+	}
+	if slot < 0 {
+		log.Log.Object(vmi).Reason(err).Error("All USB Redir devices are being used")
+		response.WriteError(http.StatusBadRequest, err)
```
Maybe `StatusTooManyRequests` is more appropriate to this case? I'm not sure.. what do you think?

victortoso replied:
Makes sense to me, altough when I tested e.g: a 3rd device when there is only 2 usbredir slots, I did not get the actual error from neither StatusBadRequest or StatusTooManyRequests from ConsoleHandler.stream(), perhaps some other fixes are needed to propagate that error to the client?

danielBelenky commented on pkg/virt-handler/rest/console.go:
```diff
@@ -45,17 +46,75 @@ type ConsoleHandler struct {
+
+	// Flag as used even if socket does not exist as a workaround to not handling how many
+	// sockets were created.
+	handler.inUse = true
```
The use of `inUse` is not very clear to me. You have a lock on the slot, why do we need the flag as well?
Also, shouldn't you release it if this function returns? If we enter the error below, wouldn't the slot remain locked?

victortoso replied:
To idea is to handle when you have multiple usbredir channels that can be used. We need to track what are in use and what are free to be used. The flag here does that. The comment points out a workaround that I'm setting it to true before checking it the socket exists - that mean we would not check it again in the future as it would stay true.

Let me know if you see a way to avoid this workaround, that is, I would like to know how many /run/kubevirt-private/.../virt-usbredir$d were created.

I'm not sure how can I replace the inUse flag with the lock. The lock seems to be used only to avoid multiple requests to the same resource... I think this code can be improved a bit still, but I'm actually new to go.

Thanks for the review!

danielBelenky commented on pkg/virtctl/usbredir/usbredir.go:
```diff
@@ -0,0 +1,217 @@
+
+	// We will connect the local USB device using usbredirserver's TCP server
+	// to the remote VM using the websocket.
+	pipeInReader, pipeInWriter := io.Pipe()
```
Shouldn't we handle the close of writer?

victortoso replied:
I've added defer _.Close() for both writers in their goroutines.

danielBelenky commented on pkg/virtctl/usbredir/usbredir.go:
```diff
@@ -0,0 +1,217 @@
+
+		bin := USBREDIRSERVER
+		if _, err := exec.LookPath(bin); err != nil {
+			usbredirResultChan <- fmt.Errorf("could not find %s binary in $PATH", USBREDIRSERVER)
```
Maybe we can move this check to the beginning of the process and fail faster?

victortoso replied:
Done! Thanks

danielBelenky reviewed (changes requested):
Very nice! A few suggestions and questions about the PR.

fabiand commented:
Thanks for the PR!

An API note, can we do the following instead to align more with the existing API

```
 spec:
   domain:
     devices:
      clientDevices:
        - usb:
            name: smartcard
        - usb:
            name: webcam
```

virtctl can then look like

```
sudo ./cluster-up/kubectl.sh usbredir 145f:019f smartcard vmi-alpine-efi
sudo ./cluster-up/kubectl.sh usbredir 145f:420f webcam vmi-alpine-efi
```

victortoso commented:
> Thanks for the PR!
> 
> An API note, can we do the following instead to align more with the existing API
> 
> ```
>  spec:
>    domain:
>      devices:
>       clientDevices:
>         - usb:
>             name: smartcard
>         - usb:
>             name: webcam
> ```
> 
> virtctl can then look like
> 
> ```
> sudo ./cluster-up/kubectl.sh usbredir 145f:019f smartcard vmi-alpine-efi
> sudo ./cluster-up/kubectl.sh usbredir 145f:420f webcam vmi-alpine-efi

Sorry, I missed this comment. I can add it without a problem.

victortoso commented:
The current PR is now using a new binary from usbredir called usbredirect ([for now](https://gitlab.freedesktop.org/spice/usbredir/-/merge_requests/2)). It does works on Windows if UsbDK is installed. One thing that I would like to work on in near future is support to migration as well, so the usb device can keep itself connected to a VM that is being migrated. Doees it seem doable from KubeVirt POV at the moment? I can add a way to handle some pause/continue methods in the client-side if needed but some guidance on what/where should I look in KubeVirt would be great ;)

Cheers,

vladikr commented:
/test all

victortoso commented:
Hi,

So I started working on the API recommendation by @fabiand below

> ```
>  spec:
>    domain:
>      devices:
>       clientDevices:
>         - usb:
>             name: smartcard
>         - usb:
>             name: webcam
> ```

On the client side, seems easy but I'm in doubt on how to handle this properly in virt-handler.

See, as we now have name for each device, I think it is fair to use it while creating the path for fd in QEMU side, e.g: `/var/run/kubevirt-private/{vmi.ObjectMeta.UID}/virt-usbredir-{usb-name}`

When the user then runs one of the following, virt-handler can use the appropriate fd if that's not yet being used.

> ```
> sudo ./cluster-up/kubectl.sh usbredir 145f:019f smartcard vmi-alpine-efi
> sudo ./cluster-up/kubectl.sh usbredir 145f:420f webcam vmi-alpine-efi
> ```

My issue/doubt is how the API should look like to provide the name argument. Before this change, I was using the helpers in virtctl which would get the stream interface with simple GET to `/apis/subresources.kubevirt.io/v1/namespaces/{namespace}/virtualmachineinstances/{name}/usbredir`.

* I thought of changing the GET to a POST and provide a `usb-name` as an extra parameter in the Body of the request but I also noted that there isn't any other POSTs api around, only PUT?

* I also thought in keeping the openapi path above to return a list of usb-name with some status like available or not (in use) and add a new GET one like before but ending with the usb-name to get the websocket stream to that device (like: `/apis/subresources.kubevirt.io/v1/namespaces/{namespace}/virtualmachineinstances/{name}/usbredir/{resourceId}`

Anyway, I realized that I might be overthinking so here I am. Suggestions?

victortoso commented:
Hi,

> >  spec:
> >    domain:
> >      devices:
> >       clientDevices:
> >         - usb:
> >             name: smartcard
> >         - usb:
> >             name: webcam

Does anyone see other possible fields  for the future, that could be used for usb itself? I'm just thinking.

One options that I thought it could be useful to add and implement is usb filtering [(here some spice, libvirt and qemu examples)](https://www.spice-space.org/usbredir.html).

> On the client side, seems easy but I'm in doubt on how to handle this properly in virt-handler.
> 
> See, as we now have name for each device, I think it is fair to use it while creating the path for fd in QEMU side, e.g: `/var/run/kubevirt-private/{vmi.ObjectMeta.UID}/virt-usbredir-{usb-name}`

In the last update I did, I'm ignoring the clientDevices.Usb.Name for now. I didn't quite get it right so far. I think the easiest way (not sure if right way) would be a query parameter, so

GET  `/apis/subresources.kubevirt.io/v1/namespaces/{namespace}/virtualmachineinstances/{name}/usbredir?usb-name=webcam`

Would get a websocket on the webcam device. We could also have a parameterless GET on /usbredir to return a list of available or not devices. [This branch here](https://github.com/victortoso/kubevirt/tree/usbredir-get-query-param) is where I stopped trying. I can see in the logs that virt-api receives the query parameter but this does not reach virt-handler.

> Anyway, I realized that I might be over thinking so here I am. Suggestions?

I'm keeping this as WIP while usbredirect binary from usbredir is not merged and released.

Feedback is always welcomed.

victortoso commented:
Hi, another iteration! The doubts/issues I had in the previous comments are now resolved.

As soon as [usbredir-0.9.0](https://gitlab.freedesktop.org/spice/usbredir/-/merge_requests/13) gets released, I'll be removing the WIP here :)

victortoso commented:
I've rebased and removed the WIP as [usbredir 0.9.0](https://gitlab.freedesktop.org/spice/usbredir/-/releases/usbredir-0.9.0) was released a week ago. If you are a fedora user, it should be reaching stable releases [this week](https://bodhi.fedoraproject.org/updates/?search=usbredir).

@vladikr please let me know if there is something pending from my side

victortoso commented:
> Hi, another iteration! The doubts/issues I had in the previous comments are now resolved.
> 
> As soon as [usbredir-0.9.0](https://gitlab.freedesktop.org/spice/usbredir/-/merge_requests/13) gets released, I'll be removing the WIP here :)

Now is the one mentioned in the comment https://github.com/kubevirt/kubevirt/pull/4089#issuecomment-725363192

alicefr commented on pkg/virtctl/usbredir/usbredir.go:
```diff
@@ -0,0 +1,201 @@
+/*
```
If I understand correctly this command is valid only for linux because of `usbredirect`, right? You can add the build tag `+build linux`, so the command is not compiled for windows

victortoso replied:
usbredirect can be used in windows too and there is a MSI instaler as well in the [release tag](https://gitlab.freedesktop.org/spice/usbredir/-/releases/usbredir-0.9.0). I have tested usbredirect on windows but I haven't tested virtctl usbredir on windows so it likely needs adjustments.

alicefr replied:
sorry Victor, I mean the `usbredirectserver`. That is missing on windows right?

victortoso replied:
`usbredirserver` binary is not used anymore, but indeed that one is linux only. Perhaps the const name `USBREDIRSERVER` is misleading in `pkg/virtctl/usbredir/usbredir.go`

alicefr replied:
yes, thanks! I got confused by the name and the description of the PR  :)

victortoso commented:
@rmohr @vladikr seems the tests are not being triggered. Is there anything pending on my side?

victortoso commented:
Rebased.

victortoso commented:
Rebased.

alicefr commented:
/ok-to-test

alicefr commented:
/test all

alicefr commented:
/retest

alicefr commented:
/test pull-kubevirt-unit-test

alicefr commented:
/retest

victortoso commented:
Rebased.

victortoso commented:
/retest

victortoso commented:
Friendly ping to @rmohr @vladikr @danielBelenky @fabiand 

I'm pretty sure that everyone is busy with something else but in order to set my own expectations, can I get an idea on when this could be looked at?

stu-gott commented on pkg/virt-api/rest/subresource.go:
```diff
@@ -165,6 +165,27 @@ func (app *SubresourceAPIApp) putRequestHandler(request *restful.Request, respon
+	validate := func(vmi *v1.VirtualMachineInstance) *errors.StatusError {
+		condManager := controller.NewVirtualMachineInstanceConditionManager()
+		if condManager.HasCondition(vmi, v1.VirtualMachineInstancePaused) {
+			return errors.NewConflict(v1.Resource("virtualmachineinstance"), vmi.Name, fmt.Errorf("VMI is paused"))
```
What happens if a VMI is paused after a device is connected?

victortoso replied:
That's something I didn't test before, so I gave it a try this morning for a little while. If we are talking about storage, like flash disk, it works fine like `console` or `vnc` commands from `virtctl`, that is, it stays connected to the guest when I `unpause` it. I tried with vmi-alpine-efi.

With a isoc device, like a webcam, I've used the vmi-fedora plus the necessary kernel modules and gstreamer1 tools to fetch data from the device (using a fakevideosink, for this test) while I pause it. The `usbredirect` eventually crashed, that means that the guest would not have its device when the VMI is `unpaused`.

This was somehow handled in the past in the usbredir library, I've added a [callback in the library](https://gitlab.freedesktop.org/spice/usbredir/commit/a88e197b18785d6de2322b5f26484c4130a6f2b9) itself to drop isoc packets based on the network connection (e.g: Streaming high quality video when the network can't handle the bandwith).

That said, this seems to be a fix in the usbredirect itself and I'll work on that for its next release, I hope it doesn't block this PR :)

victortoso replied:
Opened an issue for this [here](https://gitlab.freedesktop.org/spice/usbredir/-/issues/19)

stu-gott commented on pkg/virtctl/usbredir/usbredir.go:
```diff
@@ -0,0 +1,201 @@
+	usbNameArg := args[1]
+	usbdeviceArg := args[0]
+
+	glog.V(2).Infof("FIXME: Handle USB label name properly: %v", usbNameArg)
```
It looks weird to put a FIXME into a log message. Was the intent to fix this now or later?

victortoso replied:
Thanks, that was a leftover. The usb label is being handled properly, that is, if you define label named `storage` for a -usb device in the yaml file, it creates `/run/kubevirt-private/$hash/virt-usbredir-storage` and that is what is used based on user command line argument, the $name when you run `virtctl usbredir $device $name $vmi`

stu-gott reviewed (commented):
/approve

Overall I think this looks good. Just a couple of minor questions.

stu-gott commented:
/cc @davidvossel

xpivarc commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:
```diff
@@ -2027,6 +2028,43 @@ func getNumberOfPodInterfaces(spec *v1.VirtualMachineInstanceSpec) int {
+	if len(clientDevices) == 0 {
+		// That's fine as this is optional
+		return nil
+	} else if len(clientDevices) < 0 {
```
How can this happen?

xpivarc replied:
This build in function will never return `<0`.

victortoso replied:
Removed this chunk in the last rebase.
Thanks!

xpivarc commented on staging/src/kubevirt.io/client-go/api/v1/schema.go:
```diff
@@ -463,8 +463,33 @@ type Devices struct {
+}
+
+const (
+	UsbClientDeviceMaxNumberOf = 4
```
Just curious, why 4 is maximum?

victortoso replied:
That's the supported value in RHV with Spice. I think this can be increased as we could have multiple clients sharing USB with the target VMI but I'd say 4 USB devices is a good start. If we plan to increase in the future, it would be good to evaluate it in QEMU too, if there are other limitations that I'm unaware of.

iholder-redhat replied:
IMO this or some explanation should be in the code as a comment. From glancing at the code it would be hard to figure out why 4 is the maximum value. Changing it would much difficult if we don't know the consequences.

victortoso commented:
/retest

xpivarc reviewed (commented):
Is there a way to test this?

victortoso commented:
> Is there a way to test this?

Sure, besides the configuration mentioned in the first comment, you need `usbredirect` binary in your path. If you are in Fedora, you just need to install `usbredir` package, latest version is 0.10.0.

In case you prefer to build from source, just:
```
$ git clone https://gitlab.freedesktop.org/spice/usbredir.git
$ cd usbredir
$ meson . build
$ ninja -C build
$ ninja -C build install
#  Default prefix is /usr/local
```

victortoso commented on pkg/virt-api/api.go:
```diff
@@ -266,6 +266,14 @@ func (app *virtAPIApp) composeSubresources() {
+			To(subresourceApp.USBRedirRequestHandler).
+			Param(rest.NamespaceParam(subws)).
+			Param(rest.NameParam(subws)).
+			Param(rest.QueryParam(subws, "name", "name associated to the requested usb char device")).
```
I'll change later Today and update it here. The reason I thought name was okay is due the fact this is a query param but I see no problem in changing it to devicename.

victortoso replied:
Extracting the log from virt-api using `devicename` as query parameter:
```json
{ "url":"/apis/subresources.kubevirt.io/v1alpha3/namespaces/default/virtualmachineinstances/vmi-alpine-efi/usbredir?devicename=storage" }
```

victortoso commented:
/retest

victortoso commented:
Rebased

iholder-redhat commented on pkg/virt-handler/rest/console.go:
```diff
@@ -54,8 +61,60 @@ func NewConsoleHandler(podIsolationDetector isolation.PodIsolationDetector, vmiI
+
+	// For simplicity, we handle one usbredir request at the time, for all VMIs
+	// handled by virt-handler
+	t.usbredirLock.Lock()
+
```
I suggest we utilize go's `defer` and use `defer t.usbredirLock.Unlock()` here.

victortoso replied:
Right, I tried that initially but IIRC, because t.stream() doesn't return while the stream is active, the Unlock() didn't get called. The main goal of this lock was to avoid a race request to the same `virt-usbredir-$label` socket.

iholder-redhat replied:
Got you.
This is a very interesting case where `defer` is indeed problematic. I can think of 3 plausible alternatives:

1. Keep what you did.
2. Use `defer t.usbredirLock.Unlock()`. In the end of the function, do this:
    ```
	t.usbredirLock.Unlock()
	t.stream(vmi, request, response, unixSocketPath, stopChan, cleanup)
	t.usbredirLock.Lock()
    ```
    This is the safest way to go since if we return from one of the `if` branches we can be sure that the lock is released. On the other hand I would agree that this has ugliness to it and it's not super obvious. It also might hurt performance since you need to aquire and release the lock for no actual reason. But since, IMHO, writing safe code is critical I would consider this option.
3. Run `t.stream()` in a background thread, i.e. `go t.stream(...)`. I'm just not sure that the stream function can run safely in parallel. I have a feeling this is more complicated that it sounds.

Up to you :)

victortoso replied:
I think I tried (3) too, I'll double check just to be sure. I don't mind going with option (2) if you think it is safer. I'll try it out and submit the changes later.

victortoso replied:
(3) does not work as returning that function kills the stream. This is hard to workaround I think, because it is a webservice callback from cmd/virt-handler/virt-handler.go

iholder-redhat commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:
```diff
@@ -2027,6 +2028,39 @@ func getNumberOfPodInterfaces(spec *v1.VirtualMachineInstanceSpec) int {
+	if len(clientDevices) > v1.UsbClientDeviceMaxNumberOf {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueNotSupported,
+			Message: fmt.Sprintf("%s does not values above %v", field.String(), v1.UsbClientDeviceMaxNumberOf),
```
I think your error message from converter.go (`"Number of USB devices (%v) exceeds the maximum allowed (%v)"`) is better :)

iholder-redhat commented on pkg/virt-launcher/virtwrap/api/schema.go:
```diff
@@ -402,6 +402,20 @@ type Devices struct {
+	Redirs      []Redirected       `xml:"redirdev,omitempty"`
+}
+
+// Redirected describes a device to be redirected
+// See: https://libvirt.org/formatdomain.html#redirected-devices
+type Redirected struct {
```
Maybe we can name it `RedirectedDevice` to be clearer.

iholder-redhat commented on pkg/virt-launcher/virtwrap/api/schema.go:
```diff
@@ -402,6 +402,20 @@ type Devices struct {
+	Source RedirectedSource `xml:"source"`
+}
+
+type RedirectedSource struct {
```
`RedirectedDeviceSource`?

iholder-redhat commented on pkg/virt-launcher/virtwrap/converter/converter.go:
```diff
@@ -850,6 +850,44 @@ func Convert_v1_Rng_To_api_Rng(_ *v1.Rng, rng *api.Rng, c *ConverterContext) err
+	clientDevices := vmi.Spec.Domain.Devices.ClientDevices
+
+	// Default is to have USB Redirection disabled
+	if clientDevices == nil || len(clientDevices) <= 0 {
+		return false, nil
+	}
+
+	numberOfDevices := len(clientDevices)
+	redirectDevices := make([]api.Redirected, numberOfDevices)
+	if numberOfDevices > v1.UsbClientDeviceMaxNumberOf {
+		err := fmt.Errorf("Number of USB devices (%v) exceeds the maximum allowed (%v)", numberOfDevices, v1.UsbClientDeviceMaxNumberOf)
+		return false, err
+	}
+
+	// Check if name is unique
+	names := make(map[string]int)
+	for i, dev := range clientDevices {
+		if _, exists := names[dev.Usb.Name]; exists {
+			err := fmt.Errorf("Name of USB devices must be unique. The Name %v is repeated.", dev.Usb.Name)
+			return false, err
+		}
+		names[dev.Usb.Name] = i
+
```
I see that you've already implemented this validation in `vmi-create-admitter.go` (line 2031).
I think that if this validation is being done at the manifest level there is no need to re-validate it here.

victortoso replied:
It is true that we are double-checking (and testing) limits outside of converter.go scope. I'll rework this bit and probably remove two tests then:
- `"should reject usb redirection there is devices with the same name"`
- `"should not enable usb redirection when numberOfDevices > limit"`

And keeping only these two:
- `"should enable usb redirection when number of USB client devices > 0`
- `"should not enable usb redirection when numberOfDevices == 0"`

iholder-redhat reviewed (commented):
Great work! see feedback below :)

victortoso commented:
/retest

victortoso commented:
Rebased against master instead of main. Fixing..

victortoso commented:
/retest

iholder-redhat commented:
/test pull-kubevirt-e2e-k8s-1.20-sig-storage

iholder-redhat commented on pkg/virt-launcher/virtwrap/converter/converter.go:
```diff
@@ -853,6 +853,33 @@ func Convert_v1_Rng_To_api_Rng(_ *v1.Rng, rng *api.Rng, c *ConverterContext) err
 	return nil
 }
 
+func Convert_v1_Usbredir_To_api_Usbredir(vmi *v1.VirtualMachineInstance, domainDevices *api.Devices, c *ConverterContext) (bool, error) {
```
`c *ConverterContext` is an unused parameter

victortoso replied:
That's true, I think I was following the other implementations. I'll double check and remove.

iholder-redhat reviewed (commented):
Thanks very much for your work! :)
Left some comments below.

BTW, do you think there's a way we can test this at the functional level?

victortoso commented on pkg/virtctl/usbredir/usbredir.go:
```diff
@@ -0,0 +1,199 @@
+	case <-sigStopChan:
+	case err = <-streamStop:
+	case err = <-k8ResChan:
+	case err = <-usbredirResultChan:
```
Sure

victortoso commented:
I'll also look into a functional test for this. On the usbredir side, it always sends a hello message to exchange capability so we can simple wrap something like that and have QEMU to reply with its capabilities. That should be enough to see the KubeVirt configuration + data path is working.

PS: I'm working on some sort of [record-replay](https://gitlab.freedesktop.org/victortoso/usbredir/-/commits/record-and-replay) usbredir devices which should help expand functional tests in the future.

fengxidi commented:
> **这个 PR 的作用/为什么我们需要它**：
> 这个 PR 允许用户使用[usbredir](https://gitlab.freedesktop.org/spice/usbredir/)（协议、库和工具）将 USB 设备从客户端机器重定向到 VMI 。该库在 QEMU 和客户端使用 usbredirect 二进制文件，在[0.9.0 版本中](https://gitlab.freedesktop.org/spice/usbredir/-/releases/usbredir-0.9.0)引入。
> 
> usbredir 主要依赖项是 libusb，它在 Linux、Windows 和 macos 中运行良好。usbredirect 二进制文件与 GLib 有额外的依赖关系，它在 Windows 和 Linux 中都经过了测试。
> 
> 该 PR 在 vi​​rtctl 下实现了一个 usbredir 命令，以调用 usbredirect 二进制文件将本地 USB 设备重定向到远程 VMI。
> 
> API如[#4089（评论）](https://github.com/kubevirt/kubevirt/pull/4089#issuecomment-725363192)所建议的，用户需要在设备下添加clientDevices字段，每个USB都有一个唯一的名称。
> 
> ```yaml
>  规格：
>    域：
>      设备：
>        clientDevices：
>         - USB：
>             名称：智能卡
>         - USB：
>             名称：网络摄像头
> ```
> 
> 用户可以通过运行以下命令将本地 USB 设备与 USB 资源名称相关联：
> 
> ```shell
> sudo ./cluster-up/kubectl.sh usbredir 145f:019f 网络摄像头 vmi-alpine-efi
> ```
> 
> 您应该在 VM 的 dmesg 中看到：
> 
> ```
> [ 1169.202733] usb 1-1: USB disconnect, device number 2
> [ 1180.630968] usb 1-1: new high-speed USB device number 4 using xhci_hcd
> [ 1180.880803] usb 1-1: New USB device found, idVendor=145f, idProduct=019f
> [ 1180.880811] usb 1-1: New USB device strings: Mfr=1, Product=2, SerialNumber=0
> [ 1180.880816] usb 1-1: Product: Trust Webcam
> [ 1180.880819] usb 1-1: Manufacturer: Novatek
> ```
> 
> **发布说明**：
> 
> ```
> Add support to USB Redirection with usbredir
> ```
May I ask whether this PR can be redirected through the HTTP interface instead of executing virtctl usbredir on the physical machine?

victortoso commented:
> May I ask whether this PR can be redirected through the HTTP interface instead of executing virtctl usbredir on the physical machine?

I'm not sure I understand the question, which HTTP interface are you talking about?

victortoso commented:
I've update the first comment with the new simplified API that was suggested off list.

I've added some basic functional testing that checks that only if VM is properly configured that this feature would work; and mocked a usbredir handshake to validate the response from QEMU's usbredir.

I tried to extend the functional tests further but didn't succeed. Still learning about Go, the testing infrastructure and KubeVirt I guess. I wished to have one final functional test that would all 4 devices and see a failure when the user tries to add the 5th. Suggestions/help are welcome in case you feel this is a must to have :)

fengxidi commented:
> > 请问这个PR是否可以通过HTTP接口重定向而不是在物理机上执行virtctl usbredir？
> 
> 我不确定我是否理解这个问题，你说的是哪个 HTTP 接口？


I mean, through an HTTP interface, the USB is redirected to the virtual machine,

**example**
PUT /**/***/usbredir
  {
    "deviceID": "0781:5567",
    "usbName": "smartcard",
    "VMName": "centos"
  }

fengxidi commented:
> > May I ask whether this PR can be redirected through the HTTP interface instead of executing virtctl usbredir on the physical machine?
> 
> I'm not sure I understand the question, which HTTP interface are you talking about?

Transmitting USB device through HTTP interface
for example
PUT /***/***/usbredir
{
    "DeviceID":"0781:5567",
    "USBName":"smartcard",
}

In addition, during my test, when I format the USB device, an error （blk__update_request: 1/0 error, dev sdb, sector 272 op 0x0: (READ)f lags 8x8 phys_seg 1 prio class 0）

victortoso commented:
> Transmitting USB device through HTTP interface

The machine that has the USB device needs to run usbredir library. In this MR virtctl does launch `usbredirect` binary which loads the USB device and wraps it with usbredir protocol. A websocket is used for communicating with the Node.

All of this to say, if you want to implement a different way to redirect USB device you can, I'm just using what seems a good fit for KubeVirt at this moment.

> In addition, during my test, when I format the USB device, an error （blk__update_request: 1/0 error, dev sdb, sector 272 op 0x0: (READ)f lags 8x8 phys_seg 1 prio class 0）

Would you mind [filing an issue](https://gitlab.freedesktop.org/spice/usbredir/-/issues) in usbredir project? There was a regression in 0.9.0 for bulk devices which was fixed in 0.10.0 and this might be it.

iholder-redhat commented on tests/usbredir_test.go:
```diff
@@ -0,0 +1,177 @@
+						return
+					}
+					if nr == 0 && err == io.EOF {
+						log.Log.Info("zero bytes read from usbredir socket.")
```
Correct me if I'm wrong, but doesn't this mean that the test failed?
Makes sense to me to report that we read an empty message `readStop <- []byte("")`, that would cause the test to fail since this is not the expected message. This way we don't need to wait for the timeout to expire.

iholder-redhat commented on tests/usbredir_test.go:
```diff
@@ -0,0 +1,177 @@
+					// 1. Part of the return value is a qemu release version, e.g: 5.2.0 (test would break with different release!)
+					// 2. Capabilities can change over time which means the message would be different then the one hardcoded, correct nonetheless.
+					// I'm keeping the helloMessageRemote to have a proof of working example that could also be used if needed.
+					Expect(len(response)).To(Equal(len(helloMessageRemote)))
```
Further to my previous comment, you can even do something like this for more clarity:
```
Expect(response).ToNot(BeEmpty(), "response should not be empty")
Expect(len(response)).To(Equal(len(helloMessageRemote)))
```

iholder-redhat reviewed (commented):
Thanks very much for your work! :)

> I tried to extend the functional tests further but didn't succeed. Still learning Go and KubeVirt I guess. I wished to have one final functional test that would use all 4 devices and see a failure when the test tries to add the 5th. Suggestions/help are welcome in case you feel this is a must to have :)

Please feel free to ask for help! You can either ask me privately or here in PR discussion. I'm sure that a lot of folks will be happy to help and get involved.

In any case, I don't think it's a must-have for this PR. You can later have a follow-up PR that will add more functional tests.

victortoso commented on tests/usbredir_test.go:
```diff
@@ -0,0 +1,177 @@
+			tests.WaitForSuccessfulVMIStart(vmi)
+		})
+
+		FIt("should fail to connect to VMI's usbredir socket", func() {
```
Ops, leftover :) Thanks!

victortoso commented on tests/usbredir_test.go:
```diff
@@ -0,0 +1,177 @@
+
+			tests.BeforeTestCleanup()
+			vmi = tests.NewRandomVMI()
+			Expect(virtClient.RestClient().Post().Resource("virtualmachineinstances").Namespace(util.NamespaceTestDefault).Body(vmi).Do(context.Background()).Error()).To(Succeed())
```
I'll change too, thanks.

victortoso commented:
Rebased and applied changes. Note that I had some small changes due the code direction dictated by #4944

Cheers,

iholder-redhat commented:
> Rebased and applied changes. Note that I had some small changes due the code direction dictated by #4944
> 
> Cheers,

Thanks very much Victor, awesome work!
/lgtm

Since this PR was approved a long time ago and much have changed I want another reviewer to look at this before finally merging it.
/hold

victortoso commented:
/retest

rmohr commented on pkg/virt-launcher/virtwrap/converter/converter.go:
```diff
@@ -903,6 +903,33 @@ func Convert_v1_Rng_To_api_Rng(_ *v1.Rng, rng *api.Rng, c *ConverterContext) err
+	redirectDevices := make([]api.RedirectedDevice, v1.UsbClientPassthroughMaxNumberOf)
+
+	for i := 0; i < v1.UsbClientPassthroughMaxNumberOf; i++ {
+		path := fmt.Sprintf("/var/run/kubevirt-private/%s/virt-usbredir-%d", vmi.ObjectMeta.UID, i)
```
I don't think that the ` vmi.ObjectMeta.UID` needs to be part of the path. `/var/run/kubevirt-private` is already a private directory inside a specific `virt-launcher` container.

Since the PR exists for quite some time, this may have indeed been needed initially.

victortoso replied:
Looking at it now, it is indeed not a must to put it under VM's UID folder but the other subresources share the same path, including `.../virt-vnc` and `.../virt-serial0` so I'd let it stay in this UID folder for consistency or consider changing them all to another path.

rmohr replied:
Works for me to leave it consistent.

rmohr commented on pkg/virt-launcher/virtwrap/converter/converter_test.go:
```diff
@@ -1749,6 +1749,20 @@ var _ = Describe("Converter", func() {
+			v1.SetObjectDefaults_VirtualMachineInstance(vmi)
+			vmi.Spec.Domain.Devices.ClientPassthrough = &v1.ClientPassthroughDevices{}
+			domain := vmiToDomain(vmi, c)
+			Expect(len(domain.Spec.Devices.Redirs)).To(Equal(4))
```
You could also check if  the usb controller gets added.

rmohr commented on pkg/virt-api/rest/usbredir.go:
```diff
@@ -0,0 +1,34 @@
+)
+
+func (app *SubresourceAPIApp) USBRedirRequestHandler(request *restful.Request, response *restful.Response) {
+	activeConnectionMetric := apimetrics.NewActiveVNCConnection(request.PathParameter("namespace"), request.PathParameter("name"))
```
A small oversight here: We will need our own metric for  that. This would reuse the metric for VNC.

victortoso replied:
Ah, indeed. I did not pay enough attention while addressing the last rebase. Thanks!

rmohr commented on pkg/virt-handler/rest/console.go:
```diff
@@ -54,8 +61,67 @@ func NewConsoleHandler(podIsolationDetector isolation.PodIsolationDetector, vmiI
+	// Small workaround with lock in order to properly use defer
+	t.usbredirLock.Unlock()
+	t.stream(vmi, request, response, unixSocketPath, stopChan, cleanup)
+	t.usbredirLock.Lock()
```
I see  that you probably do the `Lock` here to cope with the `defer Unlock` above. I think that needs improvement. 

One way to cope with that without creating explicit helper functions with a clear scope is for instance an inline function:


```go
func a() {
    func () {
       lock.Lock()
       defer lock.Unlock()
       // do something locked synchronous
    }()
    // do something unlocked now.
}
```

rmohr commented on pkg/virt-handler/rest/console.go:
```diff
@@ -54,8 +61,67 @@ func NewConsoleHandler(podIsolationDetector isolation.PodIsolationDetector, vmiI
+
+	// Small workaround with lock in order to properly use defer
+	t.usbredirLock.Unlock()
+	t.stream(vmi, request, response, unixSocketPath, stopChan, cleanup)
```
I never noticed this, but this `cleanup` is unsafe to use. It is not deferred in `t.stream`. I think the better pattern would have been:

```go
defer cleanup()
t.stream(vmi, request,response, unixSocketPath, stopChan)
```

But we need at least a `defer cleanup()` inside the `stream` function on top. Could you change the stream function to that?

victortoso replied:
Ah, nice catch. Sure, I'll address it too

victortoso replied:
Let me know if you agree with it, commit: a4d40e4be52d02d081cf0050906a3d1e4a98cc6b

rmohr reviewed (changes requested):
Nice work!

Some minor improvements around the locking should be done and I think that `stream` is not really good at actually calling the cleanup function.

victortoso commented:
/retest

rmohr reviewed (approved):
/lgtm
/approve

We should not make use anymore of the `BeforeAll` in the e3e tests but I think it is best to eliminate them in one go from all tests.

victortoso commented:
@rmohr  @itamarh is it fine to unhold it?

rmohr commented:
/unhold

fabiand commented:
:balloon:

fengxidi commented:
Have you ever tried to perform USB redirect on "USB mass storage device" for USB redirect?
In my test: USB device is not available
I use the mount command to mount this device:
```
mount /dev/sda1 /data
mount: /dev/sda1 is not a valid block device
```
Kernel error:

`[  108.509440] scsi 7:0:0:0: Attached scsi generic sg0 type 0
[  108.520331] sd 7:0:0:0: [sda] 15149056 512-byte logical blocks: (7.75 GB/7.22 GiB)
[  108.521735] sd 7:0:0:0: [sda] Write Protect is off
[  108.521738] sd 7:0:0:0: [sda] Mode Sense: 0b 00 00 08
[  108.523084] sd 7:0:0:0: [sda] No Caching mode page found
[  108.523174] sd 7:0:0:0: [sda] Assuming drive cache: write through
[  108.533673]  sda: sda1
[  108.539104] sd 7:0:0:0: [sda] Attached SCSI removable disk
[  138.805609] usb 1-3: reset high-speed USB device number 4 using xhci_hcd
[  169.830194] usb 1-3: reset high-speed USB device number 4 using xhci_hcd
[  180.321107] usb 1-3: reset high-speed USB device number 4 using xhci_hcd
[  196.812832] usb 1-3: reset high-speed USB device number 4 using xhci_hcd
[  197.303927] usb 1-3: reset high-speed USB device number 4 using xhci_hcd
[  207.794750] usb 1-3: reset high-speed USB device number 4 using xhci_hcd
[  208.174301] sd 7:0:0:0: Device offlined - not ready after error recovery
[  208.174318] sd 7:0:0:0: [sda] FAILED Result: hostbyte=DID_ABORT driverbyte=DRIVER_OK cmd_age=99s
[  208.174322] sd 7:0:0:0: [sda] CDB: Read(10) 28 00 00 00 00 80 00 00 90 00
[  208.174325] blk_update_request: I/O error, dev sda, sector 128
[  208.174434] sd 7:0:0:0: rejecting I/O to offline device
[  208.174480] sd 7:0:0:0: killing request
  `

Icedroid commented:
@victortoso how to install usbredirect binary in centos.

victortoso commented:
> @victortoso how to install usbredirect binary in centos.

usbredirect is  binary provided by [usbredir](https://gitlab.com/redhat/centos-stream/rpms/usbredir/) package. So, just dnf/yum install it.

Note that usbredirect It was introduced in version 0.9.0 but lots improvements landed later. Last release is 0.12.0.

Feel free[ to open a bug](https://gitlab.freedesktop.org/spice/usbredir/) and let me know if you hit any issues.

xinzhongyoumeng commented:
> > @victortoso how to install usbredirect binary in centos.
> 
> usbredirect is binary provided by [usbredir](https://gitlab.com/redhat/centos-stream/rpms/usbredir/) package. So, just dnf/yum install it.
> 
> Note that usbredirect It was introduced in version 0.9.0 but lots improvements landed later. Last release is 0.12.0.
> 
> Feel free[ to open a bug](https://gitlab.freedesktop.org/spice/usbredir/) and let me know if you hit any issues.

I installed the usbredir-0.12.0-1.el8.x86_64 rpm in centos8.4, but there is no  usbredirect binary. Can you tell me more about this?

victortoso commented:
> I installed the usbredir-0.12.0-1.el8.x86_64 rpm in centos8.4, but there is no usbredirect binary. Can you tell me more about this?

`usbredir` package needs to be build with [`-Dtools=enabled` ](https://gitlab.com/redhat/centos-stream/rpms/usbredir/-/blob/c9s/usbredir.spec#L52). `usbredirect` uses glib2, it needs to add that as build dependency too.

zhonglin6666 commented:
two same USB device with same ID, virtual machine has one USB

victortoso commented:
@zhonglin6666 Please open a new issue if you think you have found a bug. You can also drop questions in [Slack or over google groups](https://kubevirt.io/community/).

> two same USB device with same ID, virtual machine has one USB

`virtctl usbredir` only redirects one USB each time. So you have to run it twice to redirect two USB devices.

zhonglin6666 commented:
> @zhonglin6666 Please open a new issue if you think you have found a bug. You can also drop questions in [Slack or over google groups](https://kubevirt.io/community/).
> 
> > two same USB device with same ID, virtual machine has one USB
> 
> `virtctl usbredir` only redirects one USB each time. So you have to run it twice to redirect two USB devices.

Thx, I found the new version kubevirt has already solved it.
```
usbredir (vendor:product)|(bus-device) (VMI)
```
using bus-device works fine.
